linked_list/linked_list.py

```diff
@@ -1,80 +1,159 @@
-        pass
+        self.head = Node(value, self.head)
+
 
+    def search(self, value):
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     # Time Complexity: ?
     # Space Complexity: ?
-    def search(self, value):
-        pass
+        current = self.head
+
+        while current:
+            if current.value == value:
+                return True
+            else:
+                current = current.next
+
+        return False
 
+    def length(self):
     # method that returns the length of the singly linked list
     # Time Complexity: ?
-    # Space Complexity: ?
-    def length(self):
-        pass
+    # Space Complexity: ?   
+        current = self.head
+        count = 0
+
+        while current:
+            current = current.next
+            count += 1
 
+        return count
+
+
+    def get_at_index(self, index):
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_at_index(self, index):
-        pass
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+        if index < 0:
+            return None
+        
+        current_index = 0
+        current = self.head
+
+        while current and current_index < index:
+            current = current.next
+            current_index += 1
+
+        if current is None:
+            return None
+        
+        return current.value
 
+    def get_last(self):
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
     # Time Complexity: ?
     # Space Complexity: ?
-    def get_last(self):
-        pass
+        current = self.head
+        if current is None:
+            return None
+        
+        while current:
+            if current.next == None:
+                break
+            current = current.next
 
+        return current.value
+
+    def add_last(self, value):
     # method that inserts a given value as a new last node in the linked list
     # Time Complexity: ?
-    # Space Complexity: ?
-    def add_last(self, value):
-        pass
+    # Space Complexity: ?    
+        current = self.head
+        if current is None:
+            self.head = Node(value, None)
+        else:
+            while current:
+                if current.next == None:
+                    break
+                current = current.next
+
+            current.next = Node(value, None)
+
 
-    # method to return the max value in the linked list
-    # returns the data value and not the node
     def find_max(self):
-        pass
+    # method to return the max value in the linked list
+    # returns the data value and not the node    
+        current = self.head
+        if current is None:
+            return None
+        max = 0

```

Comment:
What if the list only holds negative numbers?

```suggestion
        max = self.head.value
```